pg_wait_sampling.c

```diff
@@ -135,149 +136,63 @@ get_max_procs_count(void)
-		}
-		else if (!strcmp(name, "pg_wait_sampling.profile_period"))
-		{
-			profile_period_found = true;
-			var->integer.variable = &pgws_collector_hdr->profilePeriod;
-			pgws_collector_hdr->profilePeriod = 10;
-		}
-		else if (!strcmp(name, "pg_wait_sampling.profile_pid"))
-		{
-			profile_pid_found = true;
-			var->_bool.variable = &pgws_collector_hdr->profilePid;
-			pgws_collector_hdr->profilePid = true;
-		}
-		else if (!strcmp(name, "pg_wait_sampling.profile_queries"))
-		{
-			profile_queries_found = true;
-			var->_bool.variable = &pgws_collector_hdr->profileQueries;
-			pgws_collector_hdr->profileQueries = true;
-		}
-	}
-
-	if (!history_size_found)
-		DefineCustomIntVariable("pg_wait_sampling.history_size",
-				"Sets size of waits history.", NULL,
-				&pgws_collector_hdr->historySize, 5000, 100, INT_MAX,
-				PGC_SUSET, 0, shmem_int_guc_check_hook, NULL, NULL);
-
-	if (!history_period_found)
-		DefineCustomIntVariable("pg_wait_sampling.history_period",
-				"Sets period of waits history sampling.", NULL,
-				&pgws_collector_hdr->historyPeriod, 10, 1, INT_MAX,
-				PGC_SUSET, 0, shmem_int_guc_check_hook, NULL, NULL);
-
-	if (!profile_period_found)
-		DefineCustomIntVariable("pg_wait_sampling.profile_period",
-				"Sets period of waits profile sampling.", NULL,
-				&pgws_collector_hdr->profilePeriod, 10, 1, INT_MAX,
-				PGC_SUSET, 0, shmem_int_guc_check_hook, NULL, NULL);
-
-	if (!profile_pid_found)
-		DefineCustomBoolVariable("pg_wait_sampling.profile_pid",
-				"Sets whether profile should be collected per pid.", NULL,
-				&pgws_collector_hdr->profilePid, true,
-				PGC_SUSET, 0, shmem_bool_guc_check_hook, NULL, NULL);
-
-	if (!profile_queries_found)
-		DefineCustomBoolVariable("pg_wait_sampling.profile_queries",
-				"Sets whether profile should be collected per query.", NULL,
-				&pgws_collector_hdr->profileQueries, true,
-				PGC_SUSET, 0, shmem_bool_guc_check_hook, NULL, NULL);
-
-	if (history_size_found
-		|| history_period_found
-		|| profile_period_found
-		|| profile_pid_found
-		|| profile_queries_found)
-	{
-		ProcessConfigFile(PGC_SIGHUP);
-	}
+	DefineCustomIntVariable("pg_wait_sampling.max_profile_entries",
+			"Sets maximum number of entries in bounded profile table.", NULL,
+			&MaxProfileEntries, 5000, 100, INT_MAX,
+			PGC_POSTMASTER, 0, NULL, NULL, NULL);
+
+	DefineCustomIntVariable("pg_wait_sampling.history_size",
+			"Sets size for ring buffer for waits history in bytes.", NULL,
+			&HistoryBufferSize, 5000, 100, INT_MAX,
+			PGC_POSTMASTER, 0, NULL, NULL, NULL);
+
+	DefineCustomIntVariable("pg_wait_sampling.history_period",
+			"Sets period of waits history sampling in milliseconds.",
+			"0 disables history populating.",
+			&HistoryPeriod, 0, 0, INT_MAX,
+			PGC_SIGHUP, 0, NULL, NULL, NULL);
+
+	DefineCustomIntVariable("pg_wait_sampling.profile_period",
+			"Sets period of waits profile sampling in milliseconds.",
+			"0 disables profiling.",
+			&ProfilePeriod, 10, 0, INT_MAX,
```

Comment:
We must pass the GUC_UNIT_MS flag, so that time units, e.g. '10ms', are accepted.

Reply:
Thanks, good addition!